code/framework/auditing/src/main/java/io/cattle/platform/iaas/api/auditing/AuditServiceImpl.java

```diff
@@ -0,0 +1,82 @@
+    @Inject
+    JsonMapper jsonMapper;
+
+    private static final Logger log = LoggerFactory.getLogger(AuditLogsRequestHandler.class);
+
+    @Override
+    public void logRequest(ApiRequest request, Policy policy) {
+        Map<String, Object> data = new HashMap<>();
+        Map requestObject = CollectionUtils.toMap(request.getRequestObject());
+        data.put("requestObject", requestObject);
+        data.put("requestParams", CollectionUtils.toMap(request.getRequestParams()));
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+        try {
+            jsonMapper.writeValue(os, request.getResponseObject());
+            data.put("responseObject", os.toString());
+        } catch (IOException e) {
+            log.error("Failed to write response object to log.");
+            data.put("responseObject", "{\"object\":\"Failed to convert response object to json.\"}");
+        }
+        data.put("responseCode", request.getResponseCode());
+        data.put("responseContentType", request.getResponseContentType());
+        Identity user = null;
+        for (Identity identity: policy.getIdentities()){
+            if (identity.getExternalIdType().contains("user")){
+                user = identity;
+                break;
+            }
+        }
+        if (user == null && policy.getIdentities().size() == 1){
+            user = policy.getIdentities().iterator().next();
+        }
+        long runtime = ((long) request.getAttribute("requestEndTime")) - ((long)request.getAttribute("requestStartTime"));
+        String authType = (String) request.getAttribute(AccountConstants.AUTH_TYPE);
+        String resourceId =request.getResponseObject() instanceof Resource ? ((Resource) request.getResponseObject()).getId(): "1a0";
+        Long parsedResourceId = resourceId == null ? (long) 0 : Long.valueOf(ApiContext.getContext().getIdFormatter().parseId(resourceId));
+        String resourceType = request.getType();
+        String eventType = resourceType + "." + (requestObject.get("action") != null ? (String) requestObject.get
+                ("action") :request.getMethod());
+        auditLogDao.create(resourceType, parsedResourceId, data, user,
+                policy.getAuthenticatedAsAccountId(), policy.getAccountId(),
+                eventType, authType, runtime, null);
+    }
+
+    @Override
+    public void logResourceModification(Resource resource, Map<String, Object> data, String eventType, String description) {
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+        try {
+            jsonMapper.writeValue(os, resource);
+            data.put("resource", os.toString());
+        } catch (IOException e) {
+            log.error("Failed to write response object to log.");
+            data.put("resource", "{\"object\":\"Failed to convert response object to json.\"}");
+        }
+        auditLogDao.create(resource.getType(), resource.getId() != null ? Long.valueOf(resource.getId()): (long) 0,
```

Comment:
To clarify: resourceId would be null for Create calls? Do you think it would be cleaner if you allow to pass null values to the auditLogDao.create, and then convert null to 0L there? So you don't have to do it in all the methods calling  auditLogDao.create